### PR TITLE
Unsafe recover

### PR DESCRIPTION
**Pull request by v01dstar**
### What problem does this PR solve?

Support online recovery for Raft majority failure https://github.com/tikv/tikv/issues/10483

### What is changed and how it works?

1. Several unsafe recover related PD commands are added.  They are: `unsafe remove-failed-store <store id>,<store id>...`, `unsafe remove-failed-store show` and `unsafe remove-failed-store history`
2. After PD received unsafe recovery request, it notifies all stores through StoreHeartbeatResponse to send a detailed report (local region and raft info) in the next heartbeat.
3. Once PD have all the reports, it generates a recover plan and sends them to each individual store (using StoreHeartbeatResponse as well) to execute.
4. The recover is considered finished if all the target states are met (by comparing following store reports with recovery plan).

### Check List

Tests

- Unit test
- Integration test
- Manual test (add detailed scripts or steps below)
- No code

Code changes

- Has configuration change: **No**
- Has HTTP API interfaces change (Don't forget to [add the declarative for API](https://github.com/tikv/pd/blob/master/docs/development.md#updating-api-documentation)): **Yes**
- Has persistent data change: **Yes**, this feature may overwrite local region, raft meta data 

Side effects

- Possible performance regression: **Not applicable**
- Increased code complexity: **Limited**
- Breaking backward compatibility: **Not applicable**

Related changes

- PR to update [`pingcap/docs`](https://github.com/pingcap/docs)/[`pingcap/docs-cn`](https://github.com/pingcap/docs-cn): WIP
- PR to update [`pingcap/tidb-ansible`](https://github.com/pingcap/tidb-ansible): WIP
- Need to cherry-pick to the release branch: **N/A**

### Release note

```release-note
Support online recovery for Raft majority failure
```

### CODE REVIEW

File: server/cluster/unsafe_recovery_controller.go [hunk 1]
```diff
@@ -0,0 +1,517 @@
...
+				status = append(status, strconv.FormatUint(storeID, 10)+"not yet applied, last report: "+getStoreDigest(u.executionReports[storeID]))
+			}
+		}
+		return status
+	}
+	return []string{"Undefined status"}
+}
+
+// History returns the history logs of the current unsafe recover operation.
+func (u *unsafeRecoveryController) History() []string {
```
[the remark this answers is not in the document]
Comment (v01dstar) [reply server/cluster/unsafe_recovery_controller.go:470]: [Example output of the cmd](https://drive.google.com/file/d/171VCj85hv4l21oRUddfRgEAG2_fBqjvp/view).

`Show` and `history` are kind of similar, they both print some status info of the recovery, while `show` only shows the brief, the current phase status, and `history` shows all the store reports,  recovery plans, and execution  results, whichever it has when the cmd is called.

File: tests/server/tso/consistency_test.go [hunk 2]
```diff
@@ -12,7 +12,6 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-//go:build tso_full_test || tso_consistency_test
```
[the remark this answers is not in the document]
Comment (v01dstar) [reply tests/server/tso/consistency_test.go:15]: That was a mistake, revert it back.

File: server/cluster/unsafe_recovery_controller.go [hunk 3]
```diff
@@ -0,0 +1,517 @@
...
+	}
+	for store, plan := range u.storeRecoveryPlans {
+		log.Info("Store plan", zap.String("store", strconv.FormatUint(store, 10)), zap.String("plan", proto.MarshalTextString(plan)))
+	}
+	u.stage = recovering
+}
+
+func getPeerDigest(peer *metapb.Peer) string {
+	return strconv.FormatUint(peer.Id, 10) + ", " + strconv.FormatUint(peer.StoreId, 10) + ", " + peer.Role.String()
+}
```
[the remark this answers is not in the document]
Comment (v01dstar) [reply server/cluster/unsafe_recovery_controller.go:392]: Done

**Review by Connor1996 — APPROVED** [review]
LGTM

File: server/cluster/unsafe_recovery_controller.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (Connor1996) [reply server/cluster/unsafe_recovery_controller.go:484]: Decide to make this feature to be experimental, let's leave the better format for the future work

File: server/cluster/unsafe_recovery_controller.go [hunk 4]
```diff
@@ -0,0 +1,400 @@
...
+
+func (u *unsafeRecoveryController) canElectLeader(region *metapb.Region) bool {
+	numFailedPeers := 0
+	for _, peer := range region.Peers {
+		if _, ok := u.failedStores[peer.StoreId]; ok {
+			numFailedPeers += 1
+		}
+	}
+	return numFailedPeers*2 < len(region.Peers)
```
[the remark this answers is not in the document]
Comment (v01dstar) [reply server/cluster/unsafe_recovery_controller.go]: Done

File: tests/pdctl/store/store_test.go [hunk 5]
```diff
@@ -357,4 +357,15 @@ func (s *storeTestSuite) TestStore(c *C) {
 	output, err = pdctl.ExecuteCommand(cmd, args...)
 	c.Assert(err, IsNil)
 	c.Assert(strings.Contains(string(output), "rate should less than"), IsTrue)
+
+	// store unsafe-recover
+	args = []string{"-u", pdAddr, "store", "unsafe-recover", "s1,s2,s3"}
```
Comment (rleungx) [inline tests/pdctl/store/store_test.go:362]: Do we also support store unsafe-recover?

File: server/api/router.go [hunk 6]
```diff
@@ -265,6 +265,15 @@ func createRouter(prefix string, svr *server.Server) *mux.Router {
 	apiRouter.HandleFunc("/gc/safepoint", serviceGCSafepointHandler.List).Methods("GET")
 	apiRouter.HandleFunc("/gc/safepoint/{service_id}", serviceGCSafepointHandler.Delete).Methods("DELETE")
 
+	// unsafe admin operation API
+	unsafeOperationHandler := newUnsafeOperationHandler(svr, rd)
+	clusterRouter.HandleFunc("/unsafe/remove-failed-stores",
```
Comment (rleungx) [inline server/api/router.go:270]: prefer to add admin prefix

File: tests/pdctl/store/store_test.go (hunk 5 above)
Comment (v01dstar) [reply tests/pdctl/store/store_test.go]: No, we do not, thanks for catching this. The test cases are remove.

We named the cmd to be `unsafe-recover` and under `store` parent cmd in the past, but changed it to the current name later. I forgot to delete this. Not sure why it did fail, maybe the `help` output is not considered as an error?

File: server/api/router.go (hunk 6 above)
Comment (v01dstar) [reply server/api/router.go]: DOne

**Review by rleungx — APPROVED** [review]

**Pull request merged**